### PR TITLE
feat(replays): Add filters to Network Table

### LINKED ISSUE #38011: Network Tab should be searchable
We want to be able to filter what appears in the Network table.

1. There's a dropdown for resource types, this should be a dynamic list, based on what types are available in the list. 
2. The design doesn't show it, but there should be another dropdown to filter by Status, dynamic list again based on what is in the table.
3. The search box should filter rows based on the url

IDK how we can support filtering the Duration or Timestamp columns, lets ignore that for now and get the basics above to start.

Image

Figma: https://www.figma.com/file/OPFnT04z0uy8BEUms9OU4f/Specs%3A-Details-v3?node-id=33%3A2990

### LINKED ISSUE #38326: Design Nit: DOM Events show a line under the last item when a filter is used
There shouldn't be a line to no where

Image

### PR DESCRIPTION
**Pull request by jesus4497**
### Changes
Added filters to the network table. Now the user can filter by status, resource type or search for a word that could be in the path value. 

Fixes #38011 
Fixes #38326

Sidebar layout:

https://user-images.githubusercontent.com/39612839/187786770-596965ad-0d9d-4b45-88c1-766993df4781.mp4

Topbar layout: 

https://user-images.githubusercontent.com/39612839/187787301-84c44895-99ec-432c-85f0-497d27089b69.mp4

When DOM Events are not found:

image

### Legal Boilerplate

Look, I get it. The entity doing business as "Sentry" was incorporated in the State of Delaware in 2015 as Functional Software, Inc. and is gonna need some rights from me in order to utilize my contributions in this here PR. So here's the deal: I retain all rights, title and interest in and to my contributions, and by keeping this boilerplate intact I confirm that Sentry can use, modify, copy, and redistribute my contributions, under Sentry's choice of terms.

### CODE REVIEW

File: static/app/views/replays/detail/network/index.tsx [hunk 1]
```diff
@@ -31,6 +40,15 @@ function NetworkList({replayRecord, networkSpans}: Props) {
...
   });
+  const [searchTerm, setSearchTerm] = useState('');
+  const [filters, setFilters] = useState<Filters>({});
+
+  const filteredNetworkSpans = useMemo(
+    () => getFilteredNetworkSpans(networkSpans, searchTerm, filters),
+    [filters, networkSpans, searchTerm]
+  );
+
+  const handleSearch = debounce(query => setSearchTerm(query), 150);
```
Comment (danecando) [inline static/app/views/replays/detail/network/index.tsx:51]: Debouncing probably isn't _super_ important if the query isn't triggering a network request but still could be useful to save renders. Are you sure this is working though? The function triggers a re-render and recreates the debounced function each time. Might need to memo it

File: static/app/views/replays/detail/network/utils.tsx [hunk 2]
```diff
@@ -34,3 +40,44 @@ export function sortNetwork(
...
+  networkSpans: NetworkSpan[],
+  searchTerm: string,
+  filters: Filters
+) => {
+  if (!searchTerm && Object.keys(filters).length === 0) {
+    return networkSpans;
+  }
+  return networkSpans.filter(networkSpan => {
+    const normalizedSearchTerm = searchTerm.toLowerCase();
+    const doesMatch = JSON.stringify(networkSpan.description)
```
Comment (danecando) [inline static/app/views/replays/detail/network/utils.tsx:68]: Curious why the stringify when description is typed as a string already

File: static/app/views/replays/detail/network/utils.tsx [hunk 3]
```diff
@@ -34,3 +40,44 @@ export function sortNetwork(
     return valueB > valueA ? 1 : -1;
   });
 }
+
+export const getResourceTypes = (networkSpans: NetworkSpan[]) =>
```
Comment (ryan953) [inline static/app/views/replays/detail/network/utils.tsx:44]: can we sort this and the status types, so they display in a predictable order for users

File: static/app/views/replays/detail/network/utils.tsx [hunk 4]
```diff
@@ -34,3 +40,44 @@ export function sortNetwork(
...
+  networkSpans: NetworkSpan[],
+  searchTerm: string,
+  filters: Filters
+) => {
+  if (!searchTerm && Object.keys(filters).length === 0) {
+    return networkSpans;
+  }
+  return networkSpans.filter(networkSpan => {
+    const normalizedSearchTerm = searchTerm.toLowerCase();
+    const doesMatch = networkSpan.description
```
Comment (ryan953) [inline static/app/views/replays/detail/network/utils.tsx:68]:
```suggestion
  }
  const normalizedSearchTerm = searchTerm.toLowerCase();
  return networkSpans.filter(networkSpan => {
    const doesMatch = networkSpan.description
```

**Comment by ryan953** [comment]
This depends on https://github.com/getsentry/sentry/pull/38224

We're going to merge the DOM Events filter first, then update this PR to use the same shared functions.

**Comment by Jesse-Box** [comment]
Two things I noticed @jesus4497 

* The selectors for `Resource Type` & `Status` should probably have a value "Any" if no value from the their menu's were selected.
* Seems like the focus ring of the search input is getting cut off for some reason (See below)
Screen Shot 2022-08-31 at 13 46 51

**Comment by ryan953** [comment]
> * Seems like the focus ring of the search input is getting cut off for some reason (See below)

@Jesse-Box this is happening in a few spots. It's because of the scrollable areas. We need to move the padding/margin into the lower boxes so there's room for the :focus to shine.

**Comment by ryan953** [comment]
Looks like the "Any" text isn't working properly, it's disappeared in this example:
Screen Shot 2022-08-31 at 3 42 23 PM

**Comment by jesus4497** [comment]
> Looks like the "Any" text isn't working properly, it's disappeared in this example: Screen Shot 2022-08-31 at 3 42 23 PM

Fixed! Videos in the description updated

**Comment by jesus4497** [comment]
> @Jesse-Box made the comment live that the filter Title should match the column Title (ie: Resource Type vs Type)
> 
> I think also the order of the filters should match the order of the columns in the table. Status on the left, then Type next.

Changed!

image

**Comment by jesus4497** [comment]
> @Jesse-Box made the comment live that the filter Title should match the column Title (ie: Resource Type vs Type)
> 
> I think also the order of the filters should match the order of the columns in the table. Status on the left, then Type next.

Done!

image

**Review by ryan953 — APPROVED** [review]

**Review by ryan953 — APPROVED** [review]

**Pull request merged**